### ISSUE TITLE
Exclude URLs from Tracing

### ISSUE THREAD

**Issue by jakob-o** [open-telemetry/opentelemetry-java-instrumentation#1060]
**Is your feature request related to a problem? Please describe.**
As already mentioned here https://github.com/open-telemetry/opentelemetry-specification/issues/173 I'd like to be able to exclude or sample a list of URLs / URL-Patterns from instrumentation. In my case particularly to avoid generating many events from health- and liveness-checks.

**Describe the solution you'd like**
I opened the issue https://github.com/open-telemetry/opentelemetry-java/issues/1552 to discuss if / how tracing might be disabled from instrumentation.
To my knowledge there currently is no API / SDK method to disable tracing centrally on the context.
If I understood correctly a (maybe temporary) solution might be to create a non-recording / invalid span in the HTTP instrumentation, which due to the `ParentOrElse`-Sampler, would lead to ignoring child spans as well, if the request matches a URL pattern.
Any hint to where this would be architecturally appropriately implemented is highly appreciated.
Maybe in the [HttpServerTracer](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1b9b8b49b6413039493ac9d63a425ed0a7d572f4/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpServerTracer.java#L74)?

**Describe alternatives you've considered**
We already attempted to use the `otel.trace.classes.exclude` but only succeeded in completely disabling WebMvc instrumentation.

/CC @gabrielthunig @spaletta

**Comment by anuraaga** [comment]
Been trying auto instrumentation in a container lately, and was slightly annoyed myself with tracing of health check. Then a customer trying it out independently had the same feedback - it's exactly the kind of input we're hoping for in trials :) So I may actually mark this required for GA, the UX is impacted a lot with having no control of tracing by URL pattern.

**Comment by pavolloffay** [comment]
I have started looking into this. 

Is there a proposed design for this? In OpenTracing this was an instrumentation feature. The instrumentation check if the URL matches exclude pattern if yes then the span wasn't created. However if the excluded URL uses another instrumentation (or makes downstream call) that would create a span.  The question is whether we want to exclude just specific URLs or the whole trace starting at that URL.

**Comment by iNikem** [comment]
Yes, we have a proposal right in the task's description:

>If I understood correctly a (maybe temporary) solution might be to create a non-recording / invalid span in the HTTP instrumentation, which due to the ParentOrElse-Sampler, would lead to ignoring child spans as well, if the request matches a URL pattern.
Any hint to where this would be architecturally appropriately implemented is highly appreciated.
Maybe in the HttpServerTracer?

This should lead to ignoring the whole subtree starting from that `SERVER` span.

**Comment by pavolloffay** [comment]
+1 I think that is the right way to go https://github.com/open-telemetry/opentelemetry-specification/issues/173#issuecomment-698190021. It would also make sense to have a consolidated config property for this. 

@iNikem what is the API to create non-recording span? In OT there was  `sampling.priority=bool` tag that could be applied on the span builder https://github.com/opentracing/specification/blob/master/semantic_conventions.yaml#L26

**Comment by iNikem** [comment]
I think the right way is to use one of the factory methods on `io.opentelemetry.trace.DefaultSpan`.

**Comment by pavolloffay** [comment]
Yeah, I wanted to avoid having two paths of span creation.

**Comment by anuraaga** [comment]
Sorry if double-spam, I thought I had already posted this. How about we have a special `Sampler` itself configured that delegates to the default, except for when the path matches the allow list, then it's `ParentOnly`? It means we need to refactor to make sure our tracers set attributes on `Span.Builder` instead of `Span` - a little annoying but we should have been doing that already so maybe good motivation for it.

**Comment by anuraaga** [comment]
Note that `DefaultSpan` might get renamed to something that would be naming-wise not a good fit with what we want to do here https://github.com/open-telemetry/opentelemetry-specification/pull/994#pullrequestreview-495461768

**Comment by iNikem** [comment]
>refactor to make sure our tracers set attributes on Span.Builder instead of Span

Yes, we want to do that eventually.

**Comment by cemo** [comment]
Is there a way now to exclude health check traces? I checked processors but could not find any solution too.

**Comment by iNikem** [comment]
No, this functionality is not yet implemented.

**Comment by irl-segfault** [comment]
bump.  Is there any workaround here in the meantime?  Sampling of health checks is not ideal.

**Comment by iNikem** [comment]
> bump. Is there any workaround here in the meantime? Sampling of health checks is not ideal.

The only known workaround is to write custom sampler.

But I have plans to address this issue during the next month or so.

**Comment by iNikem** [comment]
The corresponding Sampler has been implemented in the [contrib repo](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/contrib-samplers/src/main/java/io/opentelemetry/contrib/samplers/RuleBasedRoutingSampler.java). It can be added to your deployment using [extension mechanism](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension). There is no immediate plans to add that sampler into this distribution, as this requires changes in Otel Specification and that requires some effort.

**Comment by cemo** [comment]
@iNikem what is the proposed way to configure these classes from environment? Is there any mechanism to bind jvm parameter to fields?

**Comment by iNikem** [comment]
> @iNikem what is the proposed way to configure these classes from environment?

There is no such way. To use Sampler from the contrib repo you have to add it to your deployment via extension.

**Comment by trask** [comment]
> > @iNikem what is the proposed way to configure these classes from environment?
> 
> There is no such way. To use Sampler from the contrib repo you have to add it to your deployment via extension.

👍

@cemo The current configuration is limited to key/value pairs, which doesn't model the sampler rules well. In the future, we're hoping to have a richer configuration file: https://github.com/open-telemetry/opentelemetry-specification/issues/1773. Then we could add experimental configuration support for this. And eventually we're hoping for these sampler rules (or something like them) to be spec'd and supported cross-language.

**Comment by gtuk** [comment]
@iNikem Could provide an example how to use the Sampler exactly? What i did so far is:

- Clone the repo opentelemetry-java-contrib
- Build the jar (opentelemetry-contrib-samplers-1.8.0-SNAPSHOT.jar)
- And tried to run it with
 ```
 java -javaagent:path/to/opentelemetry-javaagent.jar \
     -Dotel.javaagent.extensions=opentelemetry-contrib-samplers-1.8.0-SNAPSHOT.jar
     -jar myapp.jar
```

But how do i pass the rules to the agent?
In the DemoSampler https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSampler.java it's more clear as the rule/condition is directly inside the class

**Comment by trask** [comment]
hi @gtuk! `opentelemetry-contrib-samplers` isn't published as an opentelemetry javaagent extension, so you'll need to build your own extension that consumes (and programmatically configures) the rules

**Comment by RashmiRam** [comment]
@trask The extensions are working out pretty neat 🎉 . Thanks for that. 
One question though, Sampler's `shouldSample`only works with the following data to set the sampling decision.
1. parentContext
2. traceId
3. name
4. spanKind
5. attributes
6. parentLinks
If we want sampling decision to be based on either span's status or span's events, Is there any way to achieve it using Extension sampler?

**Comment by trask** [comment]
hey @RashmiRam! check out #3907, and there was recent discussion also about this: https://github.com/open-telemetry/opentelemetry-java/issues/3963#issuecomment-988460277

**Comment by RashmiRam** [comment]
Thanks @trask. That was really helpful. Yes. https://github.com/open-telemetry/opentelemetry-java/issues/3963#issuecomment-988460277 is also my exact use case. Wanted to achieve the same thing to push all errors irrespective of the sampling decision. 
Do you think it is a valid ask to have configuration in SDK to push all errors irrespective of sampling?

**Comment by trask** [comment]
> Do you think it is a valid ask to have configuration in SDK to push all errors irrespective of sampling?

the problem is that when the sampler decides not to sample a span (which happens at the start of a span), the SDK doesn't capture any telemetry on that span. this makes unsampled spans very efficient, but it makes impossible to change your mind at the end of an unsampled span.

**Comment by welshm** [comment]
Maybe there's a better place to ask for help, but I'm trying to resolve this problem with the extension workaround. When attempting to run the extension [example here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension) I get an error because I'm trying to build for Java8 as opposed to Java11 - is that a strict requirement for using extensions?

I tried to downgrade some gradle dependencies, but I'm not that familiar with Gradle myself (mostly use Maven). If there's a better place to ask questions or get help, please let me know!

**Comment by trask** [comment]
hi @welshm! try building with Java 11, it should still produce Java 8 compatible code

**Comment by xpicio** [comment]
hello, is there any news about this feature ?

**Comment by mnadeem** [comment]
Any update on this ?

**Comment by Montyroi** [comment]
Can anyone provide an example code in github link on how to exlude urls(exlude health check urls)

**Comment by mateuszrzeszutek** [comment]
Hey @Montyroi ,
This feature has not been implemented in the "core" javaagent yet. However, you can write a sampler yourself and package it in an extension. You can use the [rule based sampler from the contrib repo](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/samplers/RuleBasedRoutingSampler.java) that Nikita mentioned a couple posts earlier:

> The corresponding Sampler has been implemented in the [contrib repo](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/contrib-samplers/src/main/java/io/opentelemetry/contrib/samplers/RuleBasedRoutingSampler.java). It can be added to your deployment using [extension mechanism](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension). There is no immediate plans to add that sampler into this distribution, as this requires changes in Otel Specification and that requires some effort.

Or implement one from scratch, e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSampler.java

**Comment by surajsjarali** [comment]
Hey @mateuszrzeszutek @trask I am trying to exclude health check endpoint, even though spankKind Is INTERNAL and parentContext. is containing /actuator/health, I am able to see the health check traces how to drop health check traces?
```
if (spanKind == SpanKind.INTERNAL && parentContext.toString().contains("/actuator/health")) {
      return SamplingResult.create(SamplingDecision.DROP);
    }
```

**Comment by surajsjarali** [comment]
> Hey @Montyroi , This feature has not been implemented in the "core" javaagent yet. However, you can write a sampler yourself and package it in an extension. You can use the [rule based sampler from the contrib repo](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/samplers/RuleBasedRoutingSampler.java) that Nikita mentioned a couple posts earlier:
> 
> > The corresponding Sampler has been implemented in the [contrib repo](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/contrib-samplers/src/main/java/io/opentelemetry/contrib/samplers/RuleBasedRoutingSampler.java). It can be added to your deployment using [extension mechanism](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension). There is no immediate plans to add that sampler into this distribution, as this requires changes in Otel Specification and that requires some effort.
> 
> Or implement one from scratch, e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSampler.java

@mateuszrzeszutek how do we configure the list of rules for rule based sampling? could you provide a example.

**Comment by jack-berg** [comment]
This is has come up a number of times for our customers as well, so I've updated our [otel java agent example](https://github.com/newrelic/newrelic-opentelemetry-examples/tree/main/java/agent-nr-config) to package an extension which [uses the rule based](https://github.com/newrelic/newrelic-opentelemetry-examples/pull/178/files#diff-a484dc283e510f71fde57b2f0cfaaa15bfee31de5689f61c41e877f6d7e22cc7R30-R36) to drop requests to spring boot actuator endpoints. Maybe seeing concrete sample code will clear up any confusion.

**Comment by roadSurfer** [comment]
Thank you so much for this, @jack-berg. I got my own version of it running now.
For anyone else using Maven, seems you need to use the following in order to create an extension fat-jar with all the runtime dependecies:
```
<plugin>
    <artifactId>maven-assembly-plugin</artifactId>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>single</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <appendAssemblyId>false</appendAssemblyId>
        <descriptorRefs>
            <descriptorRef>jar-with-dependencies</descriptorRef>
        </descriptorRefs>
    </configuration>
</plugin>
```
I was unable to get it to run sucessfully when trying to add to the classpath normally.

**Comment by hadson172** [comment]
+1 any updates on the topic ?

**Comment by trask** [comment]
hi @hadson172! @jack-berg's example above is using the javaagent.jar

it's definitely not as simple as we would like it to be. at some point we would like to add configuration-based health check exclusions

**Comment by degodev** [comment]
+1 I would love to use the agent inside aws. Since the pricing is trace amount based, it is crucial to exclude actuator Endpoints.

**Comment by pingping95** [comment]
I have no choice but to use it until the expiration function is released.

**Comment by trask** [comment]
for anyone who would like to work on adding this functionality to the base otel javaagent, here's a very high-level outline:

use a yaml configuration file to dynamically configure the [RuleBasedRoutingSampler](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/samplers/src/main/java/io/opentelemetry/contrib/sampler/RuleBasedRoutingSampler.java) during startup of the javaagent

use a system property, e.g. `-Dotel.sampling.rules.config=...`, that users can use to specify the location of the yaml configuration file (later we will have a single yaml configuration file and we can consolidate, see [recently merged configuration otep](https://github.com/open-telemetry/oteps/pull/225))

check out the [jmx-metrics](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jmx-metrics/javaagent#configuration-files) and [metric view](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/incubator#view-file-configuration) yaml configuration for some inspiration

**Comment by vmaleze** [comment]
Based on @jack-berg examples, and the possibility to embed extension directly into the agent, I've created this simple [project](https://github.com/vmaleze/opentelemetry-java-ignore-monitoring-spans) that provides both the java agent and the docker auto instrumentation that will ignore "/health" and "/metrics" calls.

**Comment by turesheim** [comment]
We have similar requirements and wrote an extension that allows you to dynamically configure what sampler to use and add filtering for span exclusion/inclusion. Configuration can be done by simply changing the configuration file while the application is running, or also by using the optional REST service for configuring multiple agents. The service also exposes Prometheus compatible metrics. You might find it useful. See https://github.com/domstolene/da-otel-agent for details.

**Comment by szilaszi** [comment]
I did a work-around for this having a look at how the OpenTelemetryAutoConfiguration class builds the SdkTracerProvider where the sampler is being set for the bean in Spring.

I duplicated the code in my bean configuration after I have seen the original bean is conditional and created a custom a sampler, which I used to drop any health check related traces.

The caveat is that the actual health check traces are coming with no names to the sampler and denying all spans with no names may drop some unintended traces... For now we are just only doing a PoC for a product, so take this as it is :)

Here is the sample for this:

```
    @Bean
    SdkTracerProvider otelSdkTracerProvider(Environment environment, ObjectProvider<SpanProcessor> spanProcessors,
                                            Sampler sampler, ObjectProvider<SdkTracerProviderBuilderCustomizer> customizers) {
        String applicationName = environment.getProperty("spring.application.name", "application");
        SdkTracerProviderBuilder builder = SdkTracerProvider.builder()
                .setSampler(new HealthCheckExclusionSampler())
                .setResource(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, applicationName)));
        spanProcessors.orderedStream().forEach(builder::addSpanProcessor);
        customizers.orderedStream().forEach((customizer) -> customizer.customize(builder));
        return builder.build();
    }

    /**
     * A sampler implementation that excludes certain health check spans from being sampled.
     * Note: I ignore all spans without a name, reason being that the actuator health checks apparently have no names.
     */
    static class HealthCheckExclusionSampler implements Sampler {

        @Override
        public SamplingResult shouldSample(Context parentContext, String traceId, String name, SpanKind spanKind, Attributes attributes, List<LinkData> parentLinks) {
            if (name.contains("/actuator/health") || name.contains("grpc.health.v1.Health/Check") || name.contains("<unspecified span name>")) {
                return SamplingResult.drop();
            }
            return SamplingResult.recordAndSample();
        }

        @Override
        public String getDescription() {
            return "HealthCheckExclusionSampler";
        }
    }

}
```

**Comment by scprek** [comment]
Definitely curious in this as well. We have some services using Micronaut and it has a simple way of excluding HTTP routes https://micronaut-projects.github.io/micronaut-tracing/latest/guide/#http but we have other tech stacks too to deal with.

**Comment by azunna1** [comment]
If you're using a collector to send traces to your backends, you can filter out those spans - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md

**Comment by trask** [comment]
linking related: https://github.com/open-telemetry/oteps/pull/240

**Comment by shameekagarwal** [comment]
my architecture - 
app -> otel agent (auto instrumentation) -> otel collector -> observability backend
for now, i wanted to filter out swagger and health check calls (caused due to k8s readiness probe)

<details>
<summary>solution 1 - filter processor. issue - resulted in orphaned spans</summary>
![Screenshot from 2023-11-17 21-47-50](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/58326896/667fac74-9eac-47bb-b8a4-1b019b4b83c3)
</details>

<details>
<summary>solution 2 - tail sampling processor. till now, looks like it works</summary>
![Screenshot from 2023-11-17 19-06-40](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/58326896/e41b844b-2cdd-4f47-a017-ebbf1bcbf969)
</details>

**Comment by nedcerneckis** [comment]
Is there any update on this?

I understand the concerns raised by @iNikem about the specification being hard to change but this feature has been requested by numerous people for quite some time now.

I think this is a very important and much-needed feature for a lot of users of OpenTelemetry.

@trask Has anyone tried to contribute to the project with this feature? I would like to take a go at implementing it for the Otel Java agent :)

**Comment by nedcerneckis** [comment]
Is there any update on this?

I understand the concerns raised by @iNikem about the specification being hard to change but this feature has been requested by numerous people for quite some time now.

I think this is a very important and much-needed feature for a lot of users of OpenTelemetry.

@trask Has anyone tried to contribute to the project with this feature? I would like to take a go at implementing it for the Otel Java agent.

**Comment by trask** [comment]
hi @nedcerneckis!

> I understand the concerns raised by @iNikem about the specification being hard to change but this feature has been requested by numerous people for quite some time now.

I don't believe we're blocked by specification work, since we can add this initially as an opt-in experimental feature (bypassing the need for a specification)

> Has anyone tried to contribute to the project with this feature? I would like to take a go at implementing it for the Otel Java agent.

not yet, that would be great, check out https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1060#issuecomment-1494948191 for very high-level sketch that could fit well with existing feature set

**Comment by kenfinnigan** [comment]
For Lumigo's distribution I implemented a `SamplerCustomizer` for use with AutoConfigure, see [here](https://github.com/lumigo-io/opentelemetry-java-distro/blob/main/custom/src/main/java/io/lumigo/javaagent/SamplingConfigurer.java).

My approach was to allow urls for client and server to be filter, or separate env vars to filter urls for only client or server spans. The urls are defined as an array of regex, such as `[".*/health.*", ".*/actuator.*"]`

**Comment by nedcerneckis** [comment]
Thank you very much @trask! Much appreciated for the info.

I'm assuming this needs to be a little more advanced than just excluding certain HTTP endpoints and include other types of rules inside this YAML config file?

**Comment by EvertonSA** [comment]
> Based on @jack-berg examples, and the possibility to embed extension directly into the agent, I've created this simple [project](https://github.com/vmaleze/opentelemetry-java-ignore-monitoring-spans) that provides both the java agent and the docker auto instrumentation that will ignore `/health` and `/metrics` calls.

life savier, for me this is the easiest solution

**Comment by apischan** [comment]
I'm one who also waiting for that feature, but I solved this problem from another side (this approach is probably not for everyone).
The task was to disable actuator healthckecks `/actuator/health` on tracing UI. So here is what I did.
I've added the filter on opentelemetry-collector's side adding processor `filter` to force it to skip spans from actuator.

Here is the documentation about processors: https://opentelemetry.io/docs/collector/configuration/#processors
Here is the list of available filters: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor

In total now I have such configuration on my project (which I'll certainly would like to change on config property):
```yaml
config:
  exporters:
      ...

  service:
    pipelines:
      traces:
        receivers: [otlp]
        processors: [batch, filter/drophttp] # do not forget to add processor otherwise it will not work
        exporters: [otlp]

  processors:
    filter/drophttp:
      error_mode: ignore
      traces:
        span:
          - attributes["http.url"] == "/actuator/health"
          - attributes["url.path"] == "/actuator/health"
```

I would be appreciated to myself if I would find this comment hours before. 😄

**Comment by Kortex** [comment]
I did the same but I ended up with orphaned spans. How are you dealing with those?

**Comment by shameekagarwal** [comment]
> my architecture - app -> otel agent (auto instrumentation) -> otel collector -> observability backend for now, i wanted to filter out swagger and health check calls (caused due to k8s readiness probe)
> 
> solution 1 - filter processor. issue - resulted in orphaned spans
> solution 2 - tail sampling processor. till now, looks like it works

@Kortex, does [tail sampling](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1060#issuecomment-1816716602) not work?

**Comment by Kortex** [comment]
@shameekagarwal I did not try out tail sampling.

Would it be possible to share your configuration?

**Comment by shameekagarwal** [comment]
i linked my comment, where the code is in a collapsible

**Comment by apischan** [comment]
So I decided to abandon the `filter/drophttp` processor and after the hours of debugging I've finally end up with the following configuration (thanks to @shameekagarwal):
```yaml
  processors:
    tail_sampling:
      policies: [
        {
          name: filter_http_url,
          type: string_attribute,
          string_attribute: {
            key: http.url,
            values: [ /actuator/health, /swagger-ui.*, /v3/api-docs.*, /favicon.ico ],
            enabled_regex_matching: true,
            invert_match: true
          }
        },
        {
          name: filter_url_path,
          type: string_attribute,
          string_attribute: {
            key: url.path,
            values: [ /actuator/health, /swagger-ui.*, /v3/api-docs.*, /favicon.ico ],
            enabled_regex_matching: true,
            invert_match: true
          }
        }
      ]
```

**Comment by EvertonSA** [comment]
one comment here, tail sampling works absolutely fantastic for almost all issues here (related to actuator) but I have a SSE stream that can be open for more than 1 hour. Tail sampling can't keep up with such long trace.

**Comment by jack-berg** [comment]
Re-opening until lingering issues are resolved which prevent declarative config from being used with the otel java agent: https://github.com/open-telemetry/opentelemetry-java-contrib/pull/1440#discussion_r1761955278

**Comment by jack-berg** [comment]
This is resolved in [2.9.0](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v2.9.0), which fixes support for [declarative config](https://opentelemetry.io/docs/specs/otel/configuration/#declarative-configuration).

See https://github.com/open-telemetry/opentelemetry-java-examples/pull/492 for an end to end demonstration of this:
- a spring boot app
- using the otel java agent
- using declarative config
- packaging the [rule based routing sampler](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/samplers) as an extension
- configuring the sampler to be the parent based sampler, with the root set to the rule based sampler excluding certain routes

**Comment by jiten686** [comment]
@jack-berg Could provide an example how to use it. What I did is 
1. Download the open telemetry java agent -2.9.0 jar.
2. Run my application with following script. 
`-v /home/user/otel-javaagent-test:/opt/otel-javaagent 		
-e JAVA_TOOL_OPTIONS='-javaagent:/opt/otel-javaagent/opentelemetry-javaagent.jar -Dotel.javaagent.configuration-file=/opt/otel-javaagent/otel-agent.properties -Dotel.javaagent.extensions=/opt/otel-javaagent/opentelemetry-samplers-1.40.0-alpha.jar' \`  

3. otel-agent.properties has following properties.
`otel.service.name=service1
otel.propagators=tracecontext,baggage
otel.instrumentation.svc-web.enabled=true
otel.instrumentation.apache-httpclient.enabled=true
otel.instrumentation.okhttp.enabled=true
otel.exporter.otlp.protocol=grpc
otel.exporter.otlp.endpoint=http://**.***.**.**:****
otel.metrics.exporter=none
[sdk-config.txt](https://github.com/user-attachments/files/17507997/sdk-config.txt)

otel.logs.exporter=none
**otel.experimental.config.file=/opt/otel-javaagent/sdk-config.yml**`

4. sdk-config file is also attached in txt format 

I am not seeing any traces for any APIs.

**Comment by trask** [comment]
hi @jiten686, are you able to get the example in the examples repo to work? https://github.com/open-telemetry/opentelemetry-java-examples/tree/main/javaagent#file-configuration

**Comment by jiten686** [comment]
Hi @trask , Oh I missed it. Now It is working. Thank you. 
I have a question. Previously, I was using an older version of the Java agent and running it along with an agent-properties file containing the following properties.
```
otel.service.name=service1
otel.propagators=tracecontext,baggage
otel.instrumentation.svc-web.enabled=true
otel.instrumentation.apache-httpclient.enabled=true
otel.instrumentation.okhttp.enabled=true
otel.exporter.otlp.protocol=grpc
otel.exporter.otlp.endpoint=http://**.***.**.**:****
otel.metrics.exporter=none
otel.logs.exporter=none
```
However, as mention in above example, we need to use an sdk-config.yaml file to specify rules under rule_based_routing. Is there a way to keep all these configurations in the properties file instead of switching to sdk-config.yaml? If so, how can these configurations be added?"

**Comment by trask** [comment]
> Is there a way to keep all these configurations in the properties file instead of switching to sdk-config.yaml?

no, you'll need to migrate those to the equivalent yaml configuration

**Comment by sergeykad** [comment]
Is it possible to configure these exclusions using environment variables? 
I am using OpenTelemetry Operator for Kubernetes and it does not support declarative configuration.

**Comment by baole229801** [comment]
Hi @jiten686 , what you have changed  to make it works? I config the same as you and traces disappeared for all APIs.

**Comment by jiten686** [comment]
@baole229801 Instead of using agent properties file, add the all those configuration in sdk-config.yaml file and set env variable as  **-e OTEL_EXPERIMENTAL_CONFIG_FILE=path/to/sdk-config.yaml.**

**Comment by siegenthalerroger** [comment]
Wondering the same as @sergeykad regarding the capability of setting these exclusions with envvars due to the lack of support from otel-operator. Is there any update to the capabilities?

**Comment by trask** [comment]
if someone is able, I think the best path forward would be to contribute declarative config support to the operator

**Comment by 0x686578** [comment]
@jack-berg thank you for this solution! I was able to get it to work with the JARs of `io.opentelemetry.javaagent:opentelemetry-javaagent` version **2.15.0** and `io.opentelemetry.contrib:opentelemetry-samplers` version **1.46.0-alpha** as an extension.

However, after switching from environment variables to declarative configuration, the Java agent does not pick up the existing `traceparent` value but always starts a new trace (also if I don't use the opentelemetry-samplers extension). I also tried to explicitly add `propagators` to my configuration file (with values tracecontext & baggage), as well as through environment and system variables, but to no avail. Is this a known issue?

**Comment by trask** [comment]
hi @0x686578, can you open a new issue and post a repro? thanks!

**Comment by 0x686578** [comment]
@trask I found that I was using an incorrect syntax for the propagators in the declarative configuration file, it should be like this (so `propagator` without an `s` - also note that it's defined at the root level - and with the extra nested `composite` key):
```
propagator:
  composite:
    - tracecontext
    - baggage
```
I found this at https://github.com/open-telemetry/opentelemetry-configuration/blob/main/examples/sdk-migration-config.yaml

I'm not sure if the fact that the Java agent doesn't use defaults if `propagator` is not defined should be considered as a bug or not.

**Comment by trask** [comment]
@0x686578 thanks! it looks like we have it wrong here: https://github.com/open-telemetry/opentelemetry-java-examples/blob/main/declarative-configuration/otel-sdk-config.yaml

did you notice a different place where we have it wrong, so we can fix that too?

**Comment by 0x686578** [comment]
@trask, right. No, I haven't found other places. I do see now that it is correct at https://github.com/open-telemetry/opentelemetry-java-examples/blob/main/javaagent/sdk-config.yaml, but I just didn't know that it is a mandatory property. In the other file that I linked it also says "If omitted, a noop propagator is used." Maybe that should be reconsidered.
